### PR TITLE
bump to minimum python version to 3.10 and test 3.12

### PR DESCRIPTION
**Pull request by sdaulton**
## Motivation
Bump minimum version of python from 3.9 -> 3.10 and start testing 3.12, since scientific python is no longer supporting py3.9: https://scientific-python.org/specs/spec-0000/

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

(Write your answer here.)

## Test Plan

Circle CI, running manual nightly job: https://github.com/sdaulton/botorch/actions/runs/8638469226

### CODE REVIEW

File: .github/workflows/deploy_on_release.yml [hunk 1]
```diff
@@ -15,10 +15,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-14", "windows-latest"]
-        python-version: ["3.9", "3.11"]
-        exclude:
-          - os: "macos-14"
-            python-version: "3.9"  # not available for macos-14
+        python-version: ["3.10", "3.11", "3.12"]
```
Comment (saitcakmak) [inline .github/workflows/deploy_on_release.yml:18]: We test oldest & newest supported versions. No need to spend extra compute on 3.11.

File: .github/workflows/deploy_on_release.yml [hunk 2]
```diff
@@ -32,7 +29,7 @@ jobs:
       run: |
         pytest -ra --cov=. --cov-report term-missing
     - name: Upload coverage
-      if: ${{ runner.os == 'Linux' && matrix.python-version == 3.9 }}
+      if: ${{ runner.os == 'Linux' && matrix.python-version == 3.10 }}
       run: |
         bash <(curl -s https://codecov.io/bash)
```
Comment (saitcakmak) [inline .github/workflows/deploy_on_release.yml:35]: Github won't let me comment there but there's another python-version 3.9 on line 49

File: setup.py [hunk 3]
```diff
@@ -98,7 +98,7 @@ def read_deps_from_file(filname):
     ],
     long_description=long_description,
     long_description_content_type="text/markdown",
-    python_requires=">=3.9",
+    python_requires=">=3.10",
```
Comment (saitcakmak) [inline setup.py:101]: Avoids the need for future updates
```suggestion
    python_requires=f">={REQUIRED_MAJOR}.{REQUIRED_MINOR}",
```

**Review by saitcakmak — CHANGES_REQUESTED** [review]

**Comment by saitcakmak** [comment]
Looks like there are 3 tutorial failures. Two of them should be resolved once the Ax PR lands. The last one seems to originate from the `cma` package. It might require an upstream fix before we can land this -- or we can temporarily disable that tutorial.

**Comment by saitcakmak** [comment]
I looked into this a bit more closely. Looks like the failures happen because we're using development version of numpy 2.0 (installed with torch nightly), which changes some things with the copy kwarg.

**Comment by Balandat** [comment]
@saitcakmak we don't need to use the numpy 2.0 dev version though, right? Or do the torch nightlies depend on it? If not we should be able to install numpy <2.0 first and then the torch nightlies in CI

**Comment by saitcakmak** [comment]
I am guessing numpy 2.0 rc1 gets installed because the install command for pytorch nightly is `pip install --pre ...`. If `--pre` is not necessary, removing it could fix the issue without having to pin numpy. I recall this causing an issue with another dependency in the recent past

**Comment by Balandat** [comment]
Yeah that's definitely the reason it gets installed. But if we manually install numpy (without the `--pre` option) prior to installing torch, it shouldn't be upgrading numpy, even when installing torch with `--pre` option.

**Comment by saitcakmak** [comment]
I think removing `--pre` (if it works) is a cleaner solution than pinning `numpy`. Both in that it will not add additional commands in various places, and it won't introduce code that needs to be cleaned up later. By the time numpy 2.0 is released, we should aim to be compatible with it -- which doens't require a lot of work, as I found in https://github.com/CMA-ES/pycma/pull/251. We likely have much less `numpy` code than cma does.

**Comment by sdaulton** [comment]
removing pre did the trick.

File: .github/workflows/nightly.yml [hunk 4]
```diff
@@ -15,10 +15,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-14", "windows-latest"]
-        python-version: ["3.9", "3.11"]
-        exclude:
-        - os: "macos-14"
-          python-version: "3.9"  # not available for macos-14
+        python-version: ["3.10", "3.11", "3.12"]
```
Comment (Balandat) [inline .github/workflows/nightly.yml:18]: let's not test 3.11 here? If things work for 3.10 and 3.12 they should also work for 3.11. And having 3.11 here will cost a bunch of GHA capacity (note that macos hosts are more expensive)

Comment (Balandat) [inline .github/workflows/nightly.yml:18]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -16,10 +16,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-14", "windows-latest"]
-        python-version: ["3.9", "3.11"]
-        exclude:
-          - os: "macos-14"
-            python-version: "3.9"  # not available for macos-14
+        python-version: ["3.10", "3.11", "3.12"]
```
Comment (Balandat) [inline .github/workflows/test.yml:19]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test.yml [hunk 6]
```diff
@@ -49,10 +46,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-14", "windows-latest"]
-        python-version: ["3.9", "3.11"]
-        exclude:
-        - os: "macos-14"
-          python-version: "3.9"  # not available for macos-14
+        python-version: ["3.10", "3.11", "3.12"]
```
Comment (Balandat) [inline .github/workflows/test.yml:49]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test_stable.yml [hunk 7]
```diff
@@ -12,10 +12,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-14", "windows-latest"]
-        python-version: ["3.9", "3.11"]
-        exclude:
-          - os: "macos-14"
-            python-version: "3.9"  # not available for macos-14
+        python-version: ["3.10", "3.11", "3.12"]
```
Comment (Balandat) [inline .github/workflows/test_stable.yml:15]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test_stable.yml [hunk 8]
```diff
@@ -37,10 +34,7 @@ jobs:
...
```
Comment (Balandat) [inline .github/workflows/test_stable.yml:37]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test_stable.yml [hunk 9]
```diff
@@ -70,10 +64,7 @@ jobs:
       fail-fast: false
       matrix:
         os: ["ubuntu-latest", "macos-14"]
-        python-version: ["3.9", "3.11"]
-        exclude:
-        - os: "macos-14"
-          python-version: "3.9"  # not available for macos-14
+        python-version: ["3.10", "3.11", "3.12"]
```
Comment (Balandat) [inline .github/workflows/test_stable.yml:67]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/nightly.yml (hunk 4 above)
Comment (saitcakmak) [inline .github/workflows/nightly.yml:18]:
```suggestion
        python-version: ["3.10", "3.12"]
```

**Review by Balandat — CHANGES_REQUESTED** [review]

File: .github/workflows/reusable_tutorials.yml [hunk 10]
```diff
@@ -43,14 +43,15 @@ jobs:
...
       uses: actions/setup-python@v5
       with:
-        python-version: "3.9"
+        python-version: "3.10"
     - name: Fetch all history for all tags and branches
       # We need to do this so setuptools_scm knows how to set the BoTorch version.
       run: git fetch --prune --unshallow
     - if: ${{ !inputs.use_stable_pytorch_gpytorch }}
       name: Install latest PyTorch & GPyTorch
+      # Ax and cma are not compatible with numpy 2.0 yet
```
Comment (saitcakmak) [inline .github/workflows/reusable_tutorials.yml:52]:
```suggestion
```

File: .github/workflows/test.yml (hunk 5 above)
Comment (saitcakmak) [inline .github/workflows/test.yml:19]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test.yml (hunk 6 above)
Comment (saitcakmak) [inline .github/workflows/test.yml:49]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test_stable.yml (hunk 7 above)
Comment (saitcakmak) [inline .github/workflows/test_stable.yml:15]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test_stable.yml (hunk 8 above)
Comment (saitcakmak) [inline .github/workflows/test_stable.yml:37]:
```suggestion
        python-version: ["3.10", "3.12"]
```

File: .github/workflows/test_stable.yml (hunk 9 above)
Comment (saitcakmak) [inline .github/workflows/test_stable.yml:67]:
```suggestion
        python-version: ["3.10", "3.12"]
```

**Review by saitcakmak — APPROVED** [review]

**Pull request closed without merging**